evdns.c

```diff
@@ -1440,8 +1628,36 @@ server_port_read(struct evdns_server_port *s) {
 			    evutil_socket_error_to_string(err), err);
 			return;
 		}
-		request_parse(packet, r, s, (struct sockaddr*) &addr, addrlen);
+		request_parse(packet, r, s, (struct sockaddr*) &addr, addrlen, NULL);
+	}
+}
+
+static int
+server_send_response(struct evdns_server_port *port, struct server_request *req)
+{
+	u16 packet_size = 0;
+	struct bufferevent *bev = NULL;
+	if (req->client) {
+		bev = req->client->connection.bev;
+		EVUTIL_ASSERT(bev);
+		EVUTIL_ASSERT(req->response_len <= 65535);
+		packet_size = htons((u16)req->response_len);
+		if (bufferevent_write(bev, &packet_size, sizeof(packet_size)))
+			goto beferevent_error;
+		if (bufferevent_write(bev, (void*)req->response, req->response_len))
+			goto beferevent_error;
```

Comment:
won't two separate writes causes problems for query pipelining?

Reply:
(and AFAIR not only here, but in `evdns_request_transmit_through_tcp` too)

Reply:
I don't quite understand how that can be a problem. We use TCP stream here. So if we send 2 responses in a row it seems that they can't be mixed and they should go in order. We can of course put size of response and response to some buffer and write this buffer with one `bufferevent_write`, but I don't understand the problem with the current variant.
Could you please add some details about your thoughts. Maybe I didn't get the idea.

Reply:
You are absolutely right, my bad